extensions/wikia/AdEngine3/src/templates/fmr-rotator.js

```diff
@@ -0,0 +1,257 @@
+import { context, scrollListener, slotService, slotTweaker, utils } from '@wikia/ad-engine';
+import { universalAdPackage } from '@wikia/ad-engine/dist/ad-products';
+import { babDetection } from '../wad/bab-detection';
+import { btLoader } from '../wad/bt-loader';
+import { recRunner } from '../wad/rec-runner';
+
+const fmrPrefix = 'incontent_boxad_';
+const refreshInfo = {
+	adVisible: false,
+	heightScrolledThreshold: 10,
+	lastRefreshTime: new Date(),
+	refreshAdPosition: 0,
+	refreshDelay: 10000,
+	startPosition: 0
+};
+
+let btRec = false;
+let recSelector = null;
+let currentRecNode = null;
+
+let currentSlotName = null;
+let currentAdSlot = null;
+let rotatorListener = null;
+let recirculationElement = null;
+
+/***********************
+ * BT rec related logic
+ ***********************/
+
+/**
+ * Duplicates ad slot and injects BT rec placement
+ *
+ * @param {function} onSuccess
+ * @returns {void}
+ */
+function applyRec(onSuccess) {
+	if (!btRec) {
+		return;
+	}
+
+	if (recSelector === null) {
+		recSelector = 'div[id*="' + btLoader.getPlacementId(`${fmrPrefix}1`) + '"]';
+	}
+
+	currentRecNode = btLoader.duplicateSlot(`${fmrPrefix}1`);
+
+	if (currentRecNode) {
+		btLoader.triggerScript();
+	}
+
+	if (onSuccess) {
+		onSuccess();
+	}
+
+	currentRecNode = document.querySelector(recSelector) || currentRecNode;
+}
+
+/**
+ * Removes BT rec node from DOM
+ *
+ * @returns {void}
+ */
+function removeRecNode() {
+	const recNode = document.querySelector(recSelector) || currentRecNode;
+
+	if (recNode) {
+		recNode.style.display = 'none';
+		recNode.remove();
+	}
+}
+
+/****************************************
+ * Helpers and condition check functions
+ ****************************************/
+
+/**
+ * Checks if minimum refresh time counter is over
+ *
+ * @returns {boolean}
+ */
+function enoughTimeSinceLastRefresh() {
+	return ((new Date()) - refreshInfo.lastRefreshTime) > refreshInfo.refreshDelay;
+}
+
+/**
+ * Checks UAP is loaded to disable FRM logic in this case
+ *
+ * @returns {boolean}
+ */
+function fanTakeoverLoaded() {
+	return universalAdPackage.isFanTakeoverLoaded();
+}
+
+/**
+ * Checks and returns page navigation bar height
+ *
+ * @returns {int}
+ */
+function getNavbarHeight() {
+	const navbar = document.getElementById('globalNavigation');
+
+	return navbar ? navbar.offsetHeight : 0;
+}
+
+/**
+ * Checks if slot can be refreshed or limit is reached
+ *
+ * @returns {boolean}
+ */
+function refreshLimitAvailable() {
+	const limitAvailable = btRec || (currentAdSlot && (currentAdSlot.getConfigProperty('repeat.index') < 20));
```

Comment:
We can't read proper index from context, because AdBlock prevents proper AdEngine run, so instead of keeping the index duplicated in fmr-rotator module I decided to enable infinite rotation for AdBlock users since we don't make DFP requests anyway.